File: dsl-api/src/main/java/cbs/dsl/api/ConditionTypes.java
package cbs.dsl.api;

import io.avaje.jsonb.Json;
import lombok.AccessLevel;
import lombok.Builder;
import lombok.NoArgsConstructor;

import java.util.Map;

@NoArgsConstructor(access = AccessLevel.PRIVATE)
public class ConditionTypes {

  @Json
  @Builder(toBuilder = true)
  public record ConditionInput(Map<String, Object> params, String eventNumber)
      implements ConditionFunction.ConditionArg {

    public static ConditionInput from(Map<String, Object> params) {
      return ConditionInput.builder()
          .params(params)
          .build();
    }
  }

  @Json
  @Builder(toBuilder = true)
  public record ConditionOutput(boolean result) implements ConditionFunction.ConditionResult {
    @Override
    public boolean getValue() {
      return result;
    }
  }
}

File: dsl-api/src/main/kotlin/cbs/dsl/api/ConditionTypes.kt
package cbs.dsl.api

import io.avaje.jsonb.Json

/**
 * Consolidated condition DSL types.
 *
 * ConditionInput carries the raw parameters and event metadata for evaluating a condition.
 * ConditionOutput wraps the boolean result of the evaluation.
 */
@Json
@Suppress("UNCHECKED_CAST")
data class ConditionInput(
    val params: Map<String, Any?>,
    val eventCode: String? = null,
    val eventNumber: Long? = null,
) {

  /** Returns parameters with nulls filtered out, suitable for [TransactionContext]. */
  fun nonNullParams(): Map<String, Any> = params.filterValues { it != null } as Map<String, Any>
}

@Json data class ConditionOutput(val result: Boolean)
